### PR TITLE
`Lectures`: Add support for new file extensions in lecture unit file

### PR DESCRIPTION
**Pull request by evgeny1984**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/) and ensured that the layout is responsive.
- [x] Following the [theming guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

Currently, it is only possible to upload the attachments units in following format: ".png", ".jpg", ".jpeg", ".svg", ".pdf", ".zip"
Uploading e.g. of .docx files is not possible.

The corresponding issue is #5249.
### Description

Now lecture unit file supports additionally following formats: ".png", ".jpg", ".jpeg", ".svg", ".pdf", ".zip", ".doc", ".docx", ".csv", ".xls", ".xlsx", ".ppt", ".pptx", ".pages", ".rtf", ".numbers", ".key", ".odt", ".ods", ".odp", ".odg", ".odf".

### Steps for Testing

Prerequisites:
- 1 Instructor 

1. Log in to Artemis
2. Navigate to Course Administration
3. Create a new lecture or open an existing one
4. Scroll down and create a new Unit of type file
5. Click the button "Select file" and ensure that all file formats are supported which are listed on the UI
6. After selecting the file, click submit and ensure that the file unit was created successfully 

#### Code Review
- [ ] Review 1
- [ ] Review 2
#### Manual Tests
- [ ] Test 1
- [ ] Test 2

### Screenshots

image
image
image

### CODE REVIEW

**Comment by evgeny1984** [comment]
> Changes look good, but I have a more general question: Why even limit the allowed file extensions at all? From what I can see, Artemis itself does not really require the file to have a specific extension, as it only offers it for download and does not process them otherwise.

I think the idea is here to avoid uploading unnecessary file formats by providing a concrete, specified set of the file extensions.

File: src/main/webapp/i18n/de/lectureUnit.json
Comment (ge65cer) [inline src/main/webapp/i18n/de/lectureUnit.json:133]: Looks a bit weird in my opinion. Maybe just write “document formats” and/or add a tooltip that reveals the individual file extensions?

Screenshot 2022-07-05 at 09 34 28

File: src/main/webapp/i18n/en/lectureUnit.json
Comment (ge65cer) [inline src/main/webapp/i18n/en/lectureUnit.json:134]: Same as above

**Review by ge65cer — CHANGES_REQUESTED** [review]
I'd also like to see these changes reflected in the user interface by adding the respective icons from [Font Awesome](https://fontawesome.com/search?q=file&s=brands%2Csolid) for the new file types.

File: src/main/webapp/app/shared/http/file-uploader.service.ts
Comment (b-fein) [inline src/main/webapp/app/shared/http/file-uploader.service.ts:17]: Would it make sense to save this as a proper array of extensions and then generate both
- this string of `acceptedFileExtensions`, and
- another string used for the tooltip text from it?

Then the translation key `fileLimitations` would only contain the size limit and the shown tooltip text is combined from the generated list and the translation. That would remove a possible source of inconsistency after future changes to this list.

Comment (krusche) [reply src/main/webapp/app/shared/http/file-uploader.service.ts:17]: I agree, the file names are language independent, so we don't even need to translate them

Comment (evgeny1984) [reply src/main/webapp/app/shared/http/file-uploader.service.ts:17]: Implemented.

**Review by ge65cer — CHANGES_REQUESTED** [review]
> I'd also like to see these changes reflected in the user interface by adding the respective icons from [Font Awesome](https://fontawesome.com/search?q=file&s=brands%2Csolid) for the new file types.

See my previous change request.

**Comment by evgeny1984** [comment]
> 

I don't think, it makes sense to provide icons for each extension, as it does not have any additional value in the tooltip. It will be just more overloaded.

**Comment by evgeny1984** [comment]
> > I'd also like to see these changes reflected in the user interface by adding the respective icons from [Font Awesome](https://fontawesome.com/search?q=file&s=brands%2Csolid) for the new file types.
> 
> See my previous change request. _Edit_: Added screenshot. We currently distinguish between, e.g., images and PDFs. I'd like something similar for text documents, spreadsheets, etc.
> 
> 

Please create a new issue for this, as it's about the lecture attachment and not lecture unit of type file.
We need another issue for that, current change was only relevant for the units.

**Review by MetinLamby — DISMISSED** [review]
tested on ts2: all file extensions are visible in the info view, creation of the unit also works and downloading the file too :)

File: src/main/webapp/app/lecture/lecture-unit/lecture-unit-management/attachment-unit-form/attachment-unit-form.component.ts
Comment (stefanwaldhauser) [inline src/main/webapp/app/lecture/lecture-unit/lecture-unit-management/attachment-unit-form/attachment-unit-form.component.ts:154]: Why the same comment as the name of the method?

**Review by stefanwaldhauser — APPROVED** [review]
Works Can upload the mentioned files.

File: src/main/webapp/app/lecture/lecture-unit/lecture-unit-management/attachment-unit-form/attachment-unit-form.component.ts
Comment (Hialus) [reply src/main/webapp/app/lecture/lecture-unit/lecture-unit-management/attachment-unit-form/attachment-unit-form.component.ts:154]: I agree. This comment is currently redundant. 
You should explain what exactly the "translation strings" are.

Also it might make more sense to rename this method to "updateTranslationStrings", but I'll leave this up to you as it is more a personal preference from my side.

File: src/main/webapp/app/shared/constants/file-extensions.constants.ts
Comment (Hialus) [inline src/main/webapp/app/shared/constants/file-extensions.constants.ts:1]: Why do you need this as well as the definition in file-uploader.service.ts? This one is only used for showing the extension to the user and could be generated from the other one or both could be generated from an array if that makes it easier.

Comment (Hialus) [reply src/main/webapp/app/shared/constants/file-extensions.constants.ts:1]: Also, this String does not mention `pages-tef` as an extension, while the other one does. And the order of the file extensions is different between the two. This is exactly the reason why you should remove this duplication.

**Review by Hialus — CHANGES_REQUESTED** [review]

File: src/main/webapp/app/shared/constants/file-extensions.constants.ts
Comment (krusche) [reply src/main/webapp/app/shared/constants/file-extensions.constants.ts:1]: I agree, the inconsistency and duplication should be fixed

**Comment by krusche** [comment]
Client test and client style checks are not passing, so we cannot merge this PR

File: src/main/webapp/app/lecture/lecture-unit/lecture-unit-management/attachment-unit-form/attachment-unit-form.component.ts
Comment (evgeny1984) [reply src/main/webapp/app/lecture/lecture-unit/lecture-unit-management/attachment-unit-form/attachment-unit-form.component.ts:154]: rename the comment

File: src/main/webapp/app/shared/constants/file-extensions.constants.ts
Comment (evgeny1984) [reply src/main/webapp/app/shared/constants/file-extensions.constants.ts:1]: Fixed the inconsistency by using an array for both code places.

File: src/main/java/de/tum/in/www1/artemis/web/rest/FileResource.java
Comment (b-fein) [inline src/main/java/de/tum/in/www1/artemis/web/rest/FileResource.java:77]:
```suggestion
    private final List<String> allowedFileExtensions = List.of("png", "jpg", "jpeg", "svg", "pdf", "zip", "odt", "ods", "odp", "odg", "odf", "doc", "docx",
            "csv", "xls", "xlsx", "ppt", "pptx", "pages", "rtf", "pages-tef", "numbers", "key");
```

(Probably requires some more reformatting for the line break within the arguments list to satisfy the automatic formatter.)

File: src/main/webapp/app/shared/constants/file-extensions.constants.ts
Comment (b-fein) [inline src/main/webapp/app/shared/constants/file-extensions.constants.ts:1]:
```suggestion
// NOTE: this list has to be kept in sync with the extensions defined in FileResource.java
export const FILE_EXTENSIONS = [
```

Comment (b-fein) [inline src/main/webapp/app/shared/constants/file-extensions.constants.ts:1]: Nitpick: Could you maybe sort the elements in this list and in `FileResource.java` in the same order? That would make a quick comparison easier.

File: src/main/webapp/i18n/de/lectureUnit.json
Comment (b-fein) [inline src/main/webapp/i18n/de/lectureUnit.json:133]:
```suggestion
                "fileLimitation": "Max. Dateigröße 10 MB",
```
Is it 10 megabytes or 10 megabits? I assume the first. In any case the `m` should be uppercase. Same for the English translation string.

**Review by b-fein — COMMENTED** [review]
I’ve got a few more small comments. With the extensions in a list the code looks a lot better.

File: src/main/webapp/i18n/de/lectureUnit.json
Comment (evgeny1984) [reply src/main/webapp/i18n/de/lectureUnit.json:133]: Thanks, fixed

File: src/main/java/de/tum/in/www1/artemis/web/rest/FileResource.java
Comment (evgeny1984) [reply src/main/java/de/tum/in/www1/artemis/web/rest/FileResource.java:77]: Thank you, its done

File: src/main/webapp/app/shared/constants/file-extensions.constants.ts
Comment (evgeny1984) [reply src/main/webapp/app/shared/constants/file-extensions.constants.ts:1]: Done, thanks!

Comment (evgeny1984) [reply src/main/webapp/app/shared/constants/file-extensions.constants.ts:1]: Done 👍

File: src/main/java/de/tum/in/www1/artemis/web/rest/FileResource.java
Comment (evgeny1984) [reply src/main/java/de/tum/in/www1/artemis/web/rest/FileResource.java]: It does not work as it is an immutable collection then and causes an error on other place

**Review by b-fein — APPROVED** [review]
RE-approve code changes

**Pull request merged**
